doc/Gemm.md

```diff
@@ -0,0 +1,405 @@
+
+- `gemm_local.hpp` - Uses local memory for increased performance on platforms that have it. Supports full vectorization.
+
+- `gemm_no_local_partial_vec.hpp` - Doesn't use local memory. Only supports partial vectorization (can only vectorize in some specific cases).
+
+- `gemm_no_local_full_vec.hpp` - Doesn't use local memory. Supports full vectorization.
+
+- `gemm_interleaved.hpp` - An alternative approach to batched `GEMM` calculations where the inputs are interleaved in contiguous memory. Uses no local memory and corresponds to HWN data layout (NWH in column major, which is what `SYCL-BLAS` uses).
+
+## Relevant CMake Variables
+
+There are several CMake variables which are specific to `GEMM` :
+
+- `NAIVE_GEMM` (Default: `OFF`) - Forces the use of the naive, reference `GEMM` implementation.
+- `GEMM_VECTORIZATION_SUPPORT` (Default: `OFF`) - Enables vectorization within the `GEMM` kernels. If `OFF` it is equivalent to passing `1` for the vector size to the `Gemm` launcher.
+- `GEMM_TALL_SKINNY_SUPPORT` (Default: `ON`) - Enables optimizations for tall, skinny matrices. Not used on all targets.
+- `BLAS_MODEL_OPTIMIZATION` - Passing a machine learning model name here (`VGG_16` or `RESNET_50`) enables optimizations for the `Gemm` sizes used in these models. Only applies to the `ARM_GPU` target.
+
+## Kernel Structure
+
+Kernels are created as partial template specializations of the `Gemm` class to optimize for more specific cases or use different features (such as using local memory). The `Gemm` class has a number of template parameters, some of which are typically used for partial specialization. The class definition (along with the naive, reference implementation) is located in `gemm_ref.hpp`.
+
+There are some small member functions which do things such as determine the number of work groups required to execute each `Gemm`. However, the actual work of each kernel is done in `Gemm::eval()`, though the approach each kernel takes within this function is slightly different. 
+
+The general goals for programming a `GEMM` kernel can be summarised as follows:
+
+- Balance register pressure with loop unrolling to achieve an optimal balance of performance.
+- Using SFINAE and templates where applicable to minimise branching and keep as many things const and compile-time as possible for increased performance.
+
+Outside of the naive, reference `Gemm` all the kernels are tile based, where each instance of the kernel calculates a small portion of the overall matrix. Much of the work in the `::eval()` functions tends to be precalculating indices, offsets and other values for use in the actual computation. Because they are tile based one of the main considerations is whether the current tile is internal or external. If it's internal that means that boundary checks can be avoided which is a significant time save and performance increase.
+
+The core of the `Gemm` computation is as follows:
+
+- Loop over the K dimension.
```

Comment:
This list should probably be enumerated if it represents steps

Reply:
Fixed, thanks!